Add Support for Exporting Boundaries

rajadain opened the pull request:
## Overview

Adds two new endpoints to export boundaries: one for only approved boundaries, and one for all boundaries. Adds a button in the front-end for administrators to do so.

Also cleans up some minor issues.

Closes #216 
Closes #172 
Closes #93 
Closes #211 
Closes #184 

### Demo

![image](https://user-images.githubusercontent.com/1430060/203607187-47292472-1d2f-4430-a1d4-452f56fa021d.png)

![image](https://user-images.githubusercontent.com/1430060/203608351-d81485b4-0451-41c7-8b8c-82853c0658e4.png)

### Notes

5e6fbf2d8f6c7318bceff46209fbf3d1f3c34289 was done when I thought that the button would live in the List NavBar, but in the end it made more sense to place it on the main page itself. It can be excluded, but I think it is still valuable, so kept it in.

## Testing Instructions

- Checkout this branch and `server`
- Login as an Administrator
- Go to the Submissions List
  - [x] Ensure you see a button to export boundaries
  - [x] Ensure you are able to export all boundaries
  - [x] Ensure you are able to export approved boundaries
- Log out, and login as a Contributor
- Open this URL: http://localhost:4545/submissions/1
- ~~Select "Other Utility" in the selector dropdown~~
  - [ ] ~~Ensure the loaded page says "Azavea Test Utility" in the top-left, not "Other Utility"~~
  - [x] Ensure you do not see the utility selector dropdown
  - [x] Ensure the loaded page says "Azavea Test Utility" in the top-left
- Open this URL: http://localhost:4545/submissions
  - [x] Ensure you do see the utility selector dropdown
  - [x] Ensure you are taken to the appropriate submissions list after

## Checklist

- [ ] `fixup!` commits have been squashed
- [ ] `CHANGELOG.md` updated with summary of features or fixes, following [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) guidelines
- [ ] `README.md` updated if necessary to reflect the changes
- [ ] CI passes after rebase

mstone121 commented on src/django/api/views/export.py:
```diff
@@ -0,0 +1,144 @@
+            )
+        ) AS feature
+        FROM api_boundary b
+            INNER JOIN api_utility u ON u.id = b.utility_id
+            LEFT JOIN (
+                SELECT MAX(id) AS id, boundary_id
+                FROM api_submission
+                GROUP BY boundary_id) ls ON b.id = ls.boundary_id
+            LEFT JOIN api_submission s ON s.id = ls.id
+                LEFT JOIN api_user du ON s.created_by_id = du.id
+            LEFT JOIN api_user su ON s.submitted_by_id = su.id
+            LEFT JOIN (
+                SELECT MAX(id) AS id, submission_id
+                FROM api_review
+                GROUP BY submission_id) lr ON s.id = lr.submission_id
+            LEFT JOIN api_review r ON r.id = lr.id
+            LEFT JOIN api_user ru ON r.reviewed_by_id = ru.id
+            LEFT JOIN (
+                SELECT MAX(id) AS id, submission_id
+                FROM api_approval
+                GROUP BY submission_id) la ON s.id = la.submission_id
+            LEFT JOIN api_approval a ON a.id = la.id
+            LEFT JOIN api_user au ON a.approved_by_id = au.id
```
`LATERAL` subqueries can reference columns from earlier in the `FROM` clause. This lets us use the date to find the latest submission/review rather than the id where the order might not be guaranteed.

More info can be found in `7.2.1.5. LATERAL Subqueries` of [7.2.1. The FROM Clause](https://www.postgresql.org/docs/12/queries-table-expressions.html#QUERIES-FROM).
 
```suggestion
            INNER JOIN api_utility u ON u.id = b.utility_id
            LEFT JOIN LATERAL (
                SELECT * FROM api_submission
                WHERE boundary_id = b.id
                ORDER BY created_at DESC
                LIMIT 1
            ) s ON TRUE
            LEFT JOIN api_user du ON s.created_by_id = du.id
            LEFT JOIN api_user su ON s.submitted_by_id = su.id
            LEFT JOIN LATERAL (
                SELECT * FROM api_review
                WHERE submission_id = s.id
                ORDER BY created_at DESC
                LIMIT 1
            ) r ON TRUE
            LEFT JOIN api_user ru ON r.reviewed_by_id = ru.id
            LEFT JOIN LATERAL (
                SELECT * FROM api_approval
                WHERE submission_id = s.id
                ORDER BY approved_at DESC
                LIMIT 1
            ) a ON TRUE
            LEFT JOIN api_user au ON a.approved_by_id = au.id
```

rajadain replied:
Ah interesting! I hadn't heard of this before. Still don't quite have an intuition after reading up on it. I think of subqueries are row-level operations, and joins as table-level operations, so it's strange to have a subquery in a join statement. I timed them in psql though and the LATERAL JOIN appears to be running faster (???) which does not intuit well.

Nevertheless, the results are solid, and the reasoning makes sense, so I'm going with it.

[ed267d5](https://github.com/azavea/iow-boundary-tool/pull/218/commits/ed267d581e48ea8ff465e25a6fba7f5096d4dca8)

mstone121 reviewed (approved):
Looks good! Everything is working as expected.

It's unfortunate that Django does not have more support for database views (perhaps due to vendor differences) or else we wouldn't have to replicate status/latest logic.

rajadain commented on src/django/api/views/export.py:
```diff
@@ -0,0 +1,144 @@
+                'pwsid',           u.pwsid,
+                'city',            u.address_city,
+                'state',           u.state_id,
+                'drafted_at',      s.created_at,
+                'drafted_by',      du.email,
```
Good point. I made the all boundaries endpoint afterwards, will include them in the approved one too, and generally align the data structures.

rajadain replied:
[4a6815c](https://github.com/azavea/iow-boundary-tool/pull/218/commits/4a6815cb98a0a61c66afcf62efa7b176dc262829)

rajadain commented:
Thanks for taking a look and your suggestions! I was able to solve #184 after all, updated the testing instructions to reflect that.

I also reverted the commit fixing #172 since that is now fixed by this new work. Will end up dropping that before merging.

rajadain commented:
Would appreciate another quick look for the later additions to this PR

mstone121 commented on src/app/src/components/UtilityGuard.js:
```diff
@@ -25,6 +27,25 @@ export default function UtilityGuard({ children }) {
     const dispatch = useDispatch();
+    const { pathname } = useLocation();
+
+    let boundaryId = null;
+
+    if (pathname.startsWith('/submissions/')) {
+        boundaryId = pathname.substring('/submissions/'.length);
+        if (pathname.endsWith('/draw')) {
+            boundaryId = boundaryId.substring(
+                0,
+                boundaryId.length - '/draw'.length
+            );
+        }
+    }
```
```suggestion
    const boundaryId = pathname.match(/^\/submissions\/(\d+)/)?.[1];
```

rajadain replied:
Perfect. Fixed in [b555cbc](https://github.com/azavea/iow-boundary-tool/pull/218/commits/b555cbc95cb4f6234bde07de95fc5e8bf577a613)

mstone121 reviewed (approved):
👍 Thanks for adding the utility auto-select!

rajadain commented:
Rebased on latest `develop`, will merge once green

rajadain commented on src/app/src/components/UtilityGuard.js:
```diff
@@ -25,6 +27,15 @@ export default function UtilityGuard({ children }) {
     const dispatch = useDispatch();
+    const { pathname } = useLocation();
+
+    const boundaryId = pathname.match(/^\/submissions\/(\d+)/)?.[1];
```
We use this here instead of `useBoundaryId` because this component is above the `Route` that declares that parameter

rajadain commented:
Thanks for the multiple thorough reviews and all the great suggestions!
